### PR TITLE
Add server-side HTML no-cache headers and client-side static asset cache headers; modify deployment script to include static resources for up to 6 months

### PR DESCRIPTION
**Pull request by cshashwat**
## Description

Previously, server side HTML files did not have no-cache headers. This is now fixed.

Client-side files did have no-cache headers for HTML files but did not have cache headers for static resources; this is now fixed. Static resources are cached for up to 6 months.

To ensure safe deployment, the deploy script now deploys old resources along with new resources, if the old resources are less than 6 months old. After six months, these older resources are no longer deployed, and the browser should refresh.

## Related Issue

https://github.com/OfficeDev/script-lab/issues/521
https://github.com/OfficeDev/script-lab/issues/459

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: config/deploy.js [hunk 1]
```diff
@@ -129,6 +133,42 @@ function deployBuild(url, folder) {
...
+    let newAssetsPath = path.resolve(__dirname, 'bundles');
+
+    // Parse old history file if it exists
+    let history = {};
+    try {
+        history = JSON.parse(fs.readFileSync(oldHistoryPath).toString());
+    } catch (e) {}
+
+    // Add new asset files to history, with current timestamp
+    let newAssets = fs.readdirSync(newAssetsPath);
```
Comment (dsaunders45459) [inline config/deploy.js]: Is this an absolute path, relative path, or filename returned here? We will want to use relative path to root.

File: config/deploy.js [hunk 2]
```diff
@@ -129,6 +133,42 @@ function deployBuild(url, folder) {
...
+        history[asset] = { time: now };
+    }
+
+    try {
+        fs.accessSync(oldAssetsPath);
+        let existingAssets = fs.readdirSync(oldAssetsPath);
+        for (asset of existingAssets) {
+            let assetPath = path.resolve(newAssetsPath, asset);
+            // Check if old assets don't name-conflict and are less than six months old
+            if (!fs.existsSync(assetPath) && (now - history[asset].time < 15768000)) {
```
Comment (dsaunders45459) [inline config/deploy.js]: Should we delete the older entries from the json file as well through "delete history[asset]"

File: src/server/server.ts [hunk 3]
```diff
@@ -317,6 +318,7 @@ function compileCommon(req: express.Request, res: express.Response, wrapWithRunn
             }
 
             timer.stop();
+            res.setHeader('Cache-Control', 'no-cache, no-store');
```
Comment (dsaunders45459) [inline src/server/server.ts]: POST responses are not cacheable unless cache-control headers are set, so this may not be necessary.

Comment (dsaunders45459) [inline src/server/server.ts]: You need to add cache-control headers to the error page as well.
    const html = await generateErrorHtml(error, strings);
    return res.contentType('text/html').status(200).send(html);

File: config/deploy.js (hunk 1 above)
Comment (cshashwat) [inline config/deploy.js]: Discussed offline. These are just the filenames.

File: config/deploy.js (hunk 2 above)
Comment (cshashwat) [inline config/deploy.js]: Discussed offline. For right now, have decided to keep the history file for all files, in the event that users start running into script lab errors, we could look up their files here.

File: src/server/server.ts (hunk 3 above)
Comment (cshashwat) [inline src/server/server.ts]: Added.

**Pull request merged**